pkg/distributor/receiver/shim.go

```diff
@@ -0,0 +1,318 @@
+func New(receiverCfg map[string]interface{}, pusher BatchPusher, receiverFormat string, receiverDrainTimeout time.Duration, middleware Middleware, logLevel logging.Level) (services.Service, error) {
+	shim := &receiversShim{
+		pusher:       pusher,
+		format:       receiverFormat,
+		logger:       util_log.Logger,
+		drainTimeout: receiverDrainTimeout,
+	}
+
+	receiverFactories, err := component.MakeReceiverFactoryMap(
+		otlpreceiver.NewFactory(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	p := config.NewMapFromStringMap(map[string]interface{}{
+		"receivers": receiverCfg,
+	})
+	cfgs, err := configunmarshaler.NewDefault().Unmarshal(p, component.Factories{
+		Receivers: receiverFactories,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	params := component.ReceiverCreateSettings{TelemetrySettings: component.TelemetrySettings{
+		Logger:         zap.NewNop(),
+		TracerProvider: trace.NewNoopTracerProvider(),
+		MeterProvider:  metric.NewNoopMeterProvider(),
+	}}
+
+	for componentID, cfg := range cfgs.Receivers {
+		factoryBase := receiverFactories[componentID.Type()]
+		if factoryBase == nil {
+			return nil, fmt.Errorf("receiver factory not found for type: %s", componentID.Type())
+		}
+
+		receiver, err := factoryBase.CreateLogsReceiver(ctx, params, cfg, middleware.Wrap(shim))
+		if err != nil {
+			return nil, err
+		}
+
+		shim.receivers = append(shim.receivers, receiver)
+	}
+
+	shim.Service = services.NewIdleService(shim.starting, shim.stopping)
+
+	return shim, nil
+}
+func (r *receiversShim) starting(ctx context.Context) error {
+	for _, receiver := range r.receivers {
+		r.logger.Log("msg", "receiver start")
+		err := receiver.Start(ctx, r)
+		if err != nil {
+			return fmt.Errorf("error starting receiver %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Called after distributor is asked to stop via StopAsync.
+func (r *receiversShim) stopping(_ error) error {
+	// when shutdown is called on the receiver it immediately shuts down its connection
+	// which drops requests on the floor. at this point in the shutdown process
+	// the readiness handler is already down . so we are not receiving any more requests.
+	// sleep for 30 seconds to here to all pending requests to finish.
+	time.Sleep(r.drainTimeout)
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancelFn()
+
+	errs := make([]error, 0)
+
+	for _, receiver := range r.receivers {
+		err := receiver.Shutdown(ctx)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return multierr.Combine(errs...)
+	}
+
+	return nil
+}
+
+// ConsumeLogs implements consumer.Logs
+func (r *receiversShim) ConsumeLogs(ctx context.Context, ld pdata.Logs) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "distributor.ConsumeLogs")
+	defer span.Finish()
+	req, err := parseLog(ld, r.format)
+	if err != nil {
+		r.logger.Log("msg", "pusher failed to parse log data", "err", err)
```

Comment:
Again, please specify a log level here (and other occurrences):

```suggestion
		level.Error(r.logger).Log("msg", "pusher failed to parse log data", "err", err)
```